packages/expo-calendar/android/src/main/java/expo/modules/calendar/CalendarModule.kt

```diff
@@ -0,0 +1,979 @@
+      val calendarID = details.getString("id").toInt()
+      val updateUri = ContentUris.withAppendedId(CalendarContract.Calendars.CONTENT_URI, calendarID.toLong())
+      contentResolver.update(updateUri, calendarEvenFactory.eventValues, null, null)
+      calendarID
+    } else {
+      calendarEvenFactory.checkIfContainsRequiredKeys(
+        "name",
+        "title",
+        "source",
+        "color",
+        "accessLevel",
+        "ownerAccount"
+      )
+      val source = details.getArguments("source")
+      if (!source.containsKey("name")) {
+        throw Exception("new calendars require a `source` object with a `name`")
+      }
+      var isLocalAccount = false
+      if (source.containsKey("isLocalAccount")) {
+        isLocalAccount = source.getBoolean("isLocalAccount")
+      }
+      if (!source.containsKey("type") && !isLocalAccount) {
+        throw Exception("new calendars require a `source` object with a `type`, or `isLocalAccount`: true")
+      }
+
+      calendarEvenFactory
+        .put(CalendarContract.Calendars.ACCOUNT_NAME, source.getString("name"))
+        .put(CalendarContract.Calendars.ACCOUNT_TYPE, if (isLocalAccount) CalendarContract.ACCOUNT_TYPE_LOCAL else source.getString("type"))
+        .put(CalendarContract.Calendars.CALENDAR_COLOR, details.getInt("color"))
+        .put(CalendarContract.Calendars.CALENDAR_ACCESS_LEVEL, calAccessConstantMatchingString(details.getString("accessLevel")))
+        .put(CalendarContract.Calendars.OWNER_ACCOUNT, details.getString("ownerAccount"))
+
+
+      // end required fields
+      if (details.containsKey("timeZone")) {
+        calendarEvenFactory.put(CalendarContract.Calendars.CALENDAR_TIME_ZONE, details.getString("timeZone"))
+      }
+      if (details.containsKey("allowedReminders")) {
+        val array = details.getList("allowedReminders")
+        val values = arrayOfNulls<Int>(array.size)
+        for (i in array.indices) {
+          values[i] = reminderConstantMatchingString(array[i] as String?)
+        }
+        calendarEvenFactory.put(CalendarContract.Calendars.ALLOWED_REMINDERS, TextUtils.join(",", values))
+      }
+      if (details.containsKey("allowedAvailabilities")) {
+        val array = details.getList("allowedAvailabilities")
+        val values = arrayOfNulls<Int>(array.size)
+        for (i in array.indices) {
+          values[i] = availabilityConstantMatchingString(array[i] as String)
+        }
+        calendarEvenFactory.put(CalendarContract.Calendars.ALLOWED_AVAILABILITY, TextUtils.join(",", values))
+      }
+      if (details.containsKey("allowedAttendeeTypes")) {
+        val array = details.getList("allowedAttendeeTypes")
+        val values = arrayOfNulls<Int>(array.size)
+        for (i in array.indices) {
+          values[i] = attendeeTypeConstantMatchingString(array[i] as String)
+        }
+        calendarEvenFactory.put(CalendarContract.Calendars.ALLOWED_ATTENDEE_TYPES, TextUtils.join(",", values))
+      }
```

Comment:
They are looking very similar. Maybe we can move them to the factory implementation.